src/app/http-interceptors/auth-interceptor.ts

```diff
@@ -0,0 +1,45 @@
+import { Injectable } from '@angular/core';
+import { AccountService } from './../account/shared/account.service';
+import { HttpInterceptor, HttpRequest, HttpHandler, HttpErrorResponse } from '@angular/common/http';
+import { throwError } from 'rxjs';
+import { catchError } from 'rxjs/operators';
+
+@Injectable()
+export class AuthInterceptor implements HttpInterceptor {
+
+  constructor(private accountService: AccountService) { }
+
+  intercept(req: HttpRequest<any>, next: HttpHandler) {
+
+    // todas requisições batem no intercept para incluir o token no header
+    const token = this.accountService.getAuthorizationToken();
+    let request: HttpRequest<any> = req;
+
+    if (token && !this.accountService.isTokenExpired(token)) {
+      // O request é imutavel, ou seja, não é possível mudar nada
+      // Faço o clone para conseguir mudar as propriedades
+      // Passo o token de autenticação no header
+      request = req.clone({
+        headers: req.headers.set('Authorization', `Bearer ${token}`)
+      });
+    }
+
+    // retorno o request com o erro tratado
+    return next.handle(request)
+      .pipe(catchError(this.handleError)); // pipeline para tratar o erro rxjs
```

Comment:
o que é rxjs?

Reply:
Por definição é uma coleção que funciona de forma unidirecional, ou seja, ele emite notificações sempre que ocorre uma mudança em um de seus itens e a partir disto podemos executar uma ação. Digamos que ele resolve o mesmo problema que a versão anterior do Angular havia resolvido com o $watch, porém sem usar força bruta. Enquanto no $watch verificamos todo nosso escopo por alterações após cada $digest cycle(o que tem um grande custo na performance), com Observable esta verificação não acontece, pois para cada evento é emitida uma notificação para nosso Observable e então tratamos os dados.

https://medium.com/tableless/entendendo-rxjs-observable-com-angular-6f607a9a6a00